src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1309,7 +1316,223 @@ class Simplex_tree {
+                           , Vertex_handle                   v
+                           , Filtration_value                fil
+                           , int                             dim_max
+                           , std::vector<Simplex_handle>&    added_simplices)
+  {
+    /**
+     * In term of edges in the graph, inserting edge `[u,v]` only affects
+     * the subtree rooted at @p u.
+     *
+     * For a new node with label @p v, we first do a local expansion for
+     * computing the children of this new node, and then a standard expansion
+     * for its children.
+     * Nodes with label @p v (and their subtrees) already in the tree
+     * do not get affected.
+     *
+     * Nodes with label @p u get affected only if a Node with label @p v is in their same
+     * siblings set.
+     * We then try to insert "ponctually" @p v all over the subtree rooted
+     * at `Node(u)`. Each insertion of a Node with @p v label induces a local
+     * expansion at this Node (as explained above) and a sequence of "ponctual"
+     * insertion of `Node(v)` in the subtree rooted at sibling nodes of the new node,
+     * on its left.
+     */
+
+    static_assert(Options::link_nodes_by_label, "Options::link_nodes_by_label must be true");
+
+    if (u == v) { // Are we inserting a vertex?
+      auto res_ins = root_.members().emplace(u, Node(&root_,fil));
+      if (res_ins.second) { //if the vertex is not in the complex, insert it
+        added_simplices.push_back(res_ins.first); //no more insert in root_.members()
+        update_simplex_tree_after_node_insertion(res_ins.first);
+      }
+      return; //because the vertex is isolated, no more insertions.
+    }
+    // else, we are inserting an edge: ensure that u < v
+    if (v < u) { std::swap(u,v); }
+
+    //Note that we copy Simplex_handle (aka map iterators) in added_simplices
+    //while we are still modifying the Simplex_tree. Insertions in siblings may
+    //invalidate Simplex_handles; we take care of this fact by first doing all
+    //insertion in a Sibling, then inserting all handles in added_simplices.
+
+    //check whether vertices u and v are in the tree. If not, return an error.
+    auto sh_u = root_.members().find(u);
+    GUDHI_CHECK(sh_u != root_.members().end() &&
+          root_.members().find(v) != root_.members().end(),
+          std::invalid_argument(
+                  "Simplex_tree::insert_edge_as_flag - insert an edge whose vertices are not in the complex")
+                );
+
+    //check if the edge {u,v} is already in the complex, if true, nothing to do.
+    if (has_children(sh_u) &&
+        sh_u->second.children()->members().find(v) != sh_u->second.children()->members().end())
+    {
+      GUDHI_CHECK(false,"Simplex_tree::insert_edge_as_flag - insert an edge already in the complex");
+      return;
+    }
+
+    //upper bound on dimension
+    dimension_ = dim_max;
+    dimension_to_be_lowered_ = true;
```

Comment:
I admit, I didn't thought about the case `dimension_ >= dim_max ` as the way we use it in zigzag, this can never happen. This reminds me that I should perhaps also add some unit tests testing the behavior if different `dim_max` are use for different edges.

As I don't know in advance if `dim_max` will be reached or not, if I want to be more precise, I will have to test the dimension at several places. But then, indeed, I can make sure that `dimension_to_be_lowered_` remains true if it was true.

Reply:
Ok, I could probably use a similar trick than for `void expansion(int max_dim)`, I just have to be careful to not overwrite the dimension if it was initially higher. It would be the easiest way to keep track of the height of the recursion and therefore the maximal dimension added.

Reply:
If the cases that we can optimize for are not very relevant and would penalize (even slightly) the important cases, it may be best to keep the code as is. My comment is there to make sure that you think about it and make the choice deliberately. And if we keep the code as is, you can add a comment next to it explaining the reason.

Reply:
To explain the context in zigzag this will be used in: to build an oscillating rips filtration we create a vector of edges to add or remove in the filtration and for each edge to be added, we call this method to add the edge and all its new cofaces at the same filtration value. So, the complex is initially empty (filtration starts with the empty space) and the method is called with the same `max_dim` every time. If the user uses the "premade" method to compute persistence for zigzag oscillating rips, `dimension()` will never be called during the process. And at the end of the process, the complex is empty again. But the user could choose to call the iterator by hand and verify the dimension of the simplex tree between each addition/removal of edge (for statistics or whatever). But even in that case, `max_dim` would remain constant, as it is initialized with the iterator.

Here an idea of how it would look like if we keep track of the real dimension to minimize `dimension_to_be_lowered_` being true :
<pre><code><b>
    const auto tmp_dim = dimension_;
    auto tmp_max_dim = dimension_;</b>

    //for all siblings containing a Node labeled with u (including the root), run
    //compute_punctual_expansion
    List_max_vertex* nodes_with_label_u = nodes_by_label(u);//all Nodes with u label

    GUDHI_CHECK(nodes_with_label_u != nullptr,
                "Simplex_tree::insert_edge_as_flag - cannot find the list of Nodes with label u");

    for (auto node_as_hook : *nodes_with_label_u)
    {
      Node& node_u = static_cast<Node&>(node_as_hook); //corresponding node, has label u
      Simplex_handle sh_u = simplex_handle_from_node(node_u);
      Siblings * sib_u = self_siblings(sh_u);
      if (sib_u->members().find(v) != sib_u->members().end()) { //v is the label of a sibling of node_u
        int curr_dim = dimension(sib_u);
        if (dim_max == -1 || curr_dim < dim_max){
          if (!has_children(sh_u)) {
            //then node_u was a leaf and now has a new child Node labeled v
            //the child v is created in compute_punctual_expansion
            node_u.assign_children(new Siblings(sib_u, u));
          }
          <b>dimension_ = dim_max - curr_dim - 1;</b>
          compute_punctual_expansion(
                v,
                node_u.children(),
                fil,
                dim_max - curr_dim - 1, //>= 0 if dim_max >= 0, <0 otherwise
                added_simplices );
          <b>dimension_ = dim_max - dimension_;
          if (dimension_ > tmp_max_dim) tmp_max_dim = dimension_;</b>
        }
      }
    }<b>
    if (tmp_dim <= tmp_max_dim){
        dimension_ = tmp_max_dim;
        dimension_to_be_lowered_ = false;
    } else {
        dimension_ = tmp_dim;
    }</b>
</code></pre>
where  `dimension_`  is lowered in `compute_punctual_expansion` when calling `siblings_expansion` this way:
```C++
    if (dimension_ > k) {
      dimension_ = k;   //to keep track of the max height of the recursion tree
    }
```
This is absolutely not tested, but I would like you opinion about it. If you think it is worth the overhead or not. If yes, I would test it then.
Otherwise, I would use this instead as before just with a additional test:
```C++
    if (dimension_ < dim_max) {
      dimension_ = dim_max;
      dimension_to_be_lowered_ = true;
    } else if (dim_max == -1 && !dimension_to_be_lowered_) {
      ++dimension_;
      dimension_to_be_lowered_ = true;
    }
```
Or I can add a warning about the dimension in the description and let it as it.

Reply:
As the simple version does not work for the case `dim_max == -1`, I implemented the more exact version.

Reply:
I hope the dimension computation, which is useless for zigzags, doesn't end up taking too much time...

Reply:
I could do a benchmark with and without the update of the dimension? But the zigzag algorithm is already quite heavy by itself, I doubt that this will make a big difference. It could be more significant if someone only wants a oscillating Rips filtration without the barcode.

Reply:
No need, if you think it is likely ok, that's fine.